torch/fx/experimental/symbolic_shapes.py

```diff
@@ -399,6 +401,101 @@ def find_symbol_binding_fx_nodes(graph):
         if is_symbol_binding_fx_node(node)
     }
 
+
+# Analogous to ConvertIntSource
+@dataclass(frozen=True)
+class ConvertIntKey:
+    def __str__(self) -> str:
+        return ".__int__()"
+
+    def get(self, b: bool) -> int:
+        """Get the int value from bool"""
+        return int(b)
+
+
+@dataclass(frozen=True)
+class CallMethodKey:
+    name: str
+
+    def __str__(self) -> str:
+        return f".{self.name}()"
+
+    def get(self, o: Any) -> Any:
+        """Call the method on object"""
+        return getattr(o, self.name)()
+
+
+def compute_unbacked_bindings(shape_env, example_value):
+    """
+    After having run fake tensor propagation and producing example_value
+    result, traverse example_value looking for freshly bound unbacked
+    symbols and record their paths for later.  It is an error if
+    we have allocated an unbacked SymInt but it cannot be found in
+    example_value.  (NB: this means if you have a multi-output
+    function, you must call this on the tuple of tensor output, you
+    cannot wait!)
+    """
+    pending = set(shape_env.pending_fresh_unbacked_symbols)
+    if pending:
+        shape_env.pending_fresh_unbacked_symbols.clear()
+
+        def free_unbacked_symbols_with_path(
+            a, path
+        ) -> Dict[sympy.Symbol, pytree.KeyPath]:
+            r = {}
+            if isinstance(a, (tuple, list)):
+                for i in range(len(a)):
+                    r.update(
+                        free_unbacked_symbols_with_path(
+                            a[i], path + (pytree.SequenceKey(i),)
+                        )
+                    )
+            elif isinstance(a, torch.Tensor):
+                r.update(
+                    free_unbacked_symbols_with_path(
+                        a.size(), path + (CallMethodKey("size"),)
+                    )
+                )
+                r.update(
+                    free_unbacked_symbols_with_path(
+                        a.stride(), path + (CallMethodKey("stride"),)
+                    )
+                )
+                r.update(
+                    free_unbacked_symbols_with_path(
+                        a.storage_offset(), path + (CallMethodKey("storage_offset"),)
+                    )
+                )
+
+            # NB: Intentionally access _expr, not expr, do not want
+            # simplification!
+            elif (
+                isinstance(a, (torch.SymInt, torch.SymFloat))
+                and isinstance(s := a.node._expr, sympy.Symbol)
+                and s in pending
+            ):
+                r[s] = path
+                pending.remove(s)
+            # The annoyance here arises from the fact that SymBool is
+            # allocated by allocating a SymInt and then testing if it's equal
+            # to one.  So you have a complicated binding site logic for this.
+            elif (
```

Comment:
Actually... maybe a more basic question. When is it possible to see `SymBool` here? I think I imagined that:

(1) unbacked SymBools are only ever generated from some code that performs some conditional expression on an existing SymInt/SymFloat
(2) SymInt/SymFloat is guaranteed to be generated by either item() or custom op calls, which will already have had an example_value created for them and have been "processed" by this function before we perform any conditional checks on the unbacked syms.

Reply:
Don't need an else branch here. If we fail to detect all the unbacked symints, pending will still contain entries and it will raise at the end. In fact, we emphatically do NOT want to raise here, because regular operations could return stuff like 4 or (u0 + 1) which are not binding sites and we need to just support.

Reply:
The SymBool only ever occurs when we do an item() call; there's a preexisting hack that converts the boolean into an integer that is either 0 or 1 and then tests equality on it. Honestly, I should probably just figure out how to model this properly in sympy with a variable that represents booleans; the main problem is there is no way to tell if a sympy.Variable is a boolean or not (unlike is_integer).

Reply:
So concretely, when you see

```
x = bool_tensor.item()
```

in symbolic shapes, we actually allocate u0 (an unbacked integer), and then give x the value "u0 == 1". But at runtime, it's still a boolean we're getting out from the item call, and we need to convert it back into an integer so we have something to bind to u0.